Add JSON, hex, and keccak operators for variable operations

theref opened the pull request:
## Summary

Adds new operators to the condition lingo system for enhanced variable operations:
- **JSON operators**: `fromJson`, `toJson` for parsing/serializing JSON data
- **Hex operators**: `fromHex`, `toHex` for hex encoding/decoding
- **Hash operator**: `keccak` for keccak256 hashing

## Changes

- Added 5 new operators to `VARIABLE_OPERATIONS` in `nucypher/policy/conditions/lingo.py`
- Comprehensive unit tests for JSON, hex, and keccak operators
- JSON-API integration tests for variable operations
- ECDSA condition integration tests demonstrating real-world usage with context parameter concatenation

## Testing

- ✅ Unit tests for all new operators with various input types
- ✅ JSON-API tests validating variable operations in API conditions
- ✅ Integration tests showing practical usage with ECDSA conditions and Discord payload simulation

theref committed:
feat(conditions): add JSON, hex, and keccak operators

Add new variable operation operators for data transformation:
- toJson/fromJson: Convert between Python objects and JSON strings
- toHex/fromHex: Convert between bytes/strings and hex representation
- keccak: Compute keccak-256 hash (outputs hex string)

These operators enable conditions to transform data before comparison,
particularly useful for comparing hex representations with object data
to verify integrity.

theref committed:
test(conditions): add tests for JSON, hex, and keccak operators

Add comprehensive test coverage for new variable operations:
- Test JSON/hex round-trip conversions
- Test keccak hashing with known values
- Test string concatenation with += operator
- Test practical use case: comparing hex with object representation

All 167 variable operation tests passing.

theref committed:
test(conditions): add JSON-API tests for variable operations

Add integration tests demonstrating how to use new operators with
JSON-API conditions:
- Test converting API responses to JSON/hex
- Test combining JSONPath queries with operations
- Test hex comparison use case for data integrity verification
- Test serialization/deserialization of operations
- Test context variables in operation values

All 34 JSON-API condition tests passing.

KPrasch committed:
crafting a simulated discord payload compunded with ecdsa condition

KPrasch committed:
expands sequential condition testing to demo concatenation of context params passed into ECDSA vars.

theref committed:
Add newsfragment for PR #3664

cygnusv commented on nucypher/policy/conditions/lingo.py:
```diff
@@ -330,6 +330,18 @@ def __init__(self, operand: Condition):
+    "toHex": lambda a: (
+        HexBytes(a).hex()
+        if isinstance(a, (bytes, bytearray))
+        else HexBytes(a.encode() if isinstance(a, str) else str(a).encode()).hex()
```
```suggestion
        else HexBytes(str(a).encode()).hex()
```

derekpierre replied:
We have to be really careful here. I don't think we can just cast to string. Not everything can/should be turned into hex, for example floats, so someone passing a float to this function should raise a `TypeError`.

```python
In [7]: h = HexBytes(17)

In [8]: h.hex()
Out[8]: '0x11'.   # correct

In [9]: h = HexBytes(str(17).encode())

In [10]: h.hex()
Out[10]: '0x3137'.   # not correct
```

`HexBytes` already takes care of bytes, bytearray, and integer cases for us, *EXCEPT* the case when a string is passed that is non-hex (it expects any string to already be hex):

```python
In [24]: HexBytes("dead").hex()
Out[24]: '0xdead'

In [25]: HexBytes("deadbeat").hex()
---------------------------------------------------------------------------
Error                                     Traceback (most recent call last)
Cell In[25], line 1
----> 1 HexBytes("deadbeat").hex()

...
File ~/.local/share/virtualenvs/nucypher-porter-wF43iFf2/lib/python3.12/site-packages/hexbytes/_utils.py:54, in hexstr_to_bytes(hexstr)
     50     raise ValueError(
     51         f"hex string {padded_hex} may only contain [0-9a-fA-F] characters"
     52     )
     53 else:
---> 54     return binascii.unhexlify(ascii_hex)

Error: Non-hexadecimal digit found
```

Do we care to differentiate between an already provided hex string, and a regular string...my view is no differentiation i.e. any string provided should be converted to bytes

Note that these operator functions raise `TypeError` if invalid types are provided.

What about a function that does something like:
```python
def to_hex(value):
    try:
         if isinstance(value, str):
             value = value.encode("utf-8")
         h = HexBytes(value)
         return h.hex()
    except Exception as e:
         raise TypeError(f"Invalid value for hex conversion: {e}")
```

^ (Perhaps `Exception` can be `(TypeError, ValueError)` instead; not sure what other exceptions can arise from `HexBytes`)

cygnusv commented on tests/integration/conditions/test_ecdsa_condition_integration.py:
```diff
@@ -129,30 +136,257 @@ def test_ecdsa_in_compound_condition():
+    public_key_hex = "b853dd9f496723daf64bf2f5a886086f790df66e61d7b6f7f98a50c9e5ede8f3"
+    signature_hex = "0a12acb96843700b724f1c9dba3075a7fc482677e0c713eb6be63bfea406fb33b4715975f1062b2dff95127bd559ee23758911bd217760727dc44e7880bc6e04"
+    timestamp = "1749368683"
+    body = '{"app_permissions":"2248473465835073","application_id":"1380486651436073092","attachment_size_limit":10485760,"authorizing_integration_owners":{"0":"1380488052169769110"},"channel":{"flags":0,"guild_id":"1380488052169769110","icon_emoji":{"id":null,"name":"👋"},"id":"1380488052169769113","last_message_id":"1380528253168652328","name":"general","nsfw":false,"parent_id":"1380488052169769 ...
```
If these values are the same than in the previous test, maybe they can be defined as test constants.

cygnusv commented on tests/unit/conditions/test_variable_operation.py:
```diff
@@ -244,6 +266,128 @@ def test_overloaded_operators():
     assert result == "TACoTACoTACo!"
 
 
+def test_string_concatenation():
```
There's already a similar test above, `test_overloaded_operators()`.

cygnusv commented on tests/unit/conditions/test_variable_operation.py:
```diff
@@ -244,6 +266,128 @@ def test_overloaded_operators():
+        VariableOperation(operation="fromJson"),  # back to original dict
+    ]
+    result = VariableOperation.evaluate_operations(operations, initial)
+    assert result == initial
```
👏

derekpierre commented on nucypher/policy/conditions/lingo.py:
```diff
@@ -330,6 +330,18 @@ def __init__(self, operand: Condition):
     "int": lambda a: int(a),
     "str": lambda a: str(a),
+    # JSON conversion
+    "fromJson": lambda a: json.loads(a if isinstance(a, str) else a.decode()),
```
`json.loads` already takes care of typing for us:
```
def loads(s, *, cls=None, object_hook=None, parse_float=None,
        parse_int=None, parse_constant=None, object_pairs_hook=None, **kw):
    """Deserialize ``s`` (a ``str``, ``bytes`` or ``bytearray`` instance
    containing a JSON document) to a Python object.
```

```suggestion
    "fromJson": lambda a: json.loads(a),
```

derekpierre commented on tests/integration/conditions/test_ecdsa_condition_integration.py:
```diff
@@ -129,30 +136,257 @@ def test_ecdsa_in_compound_condition():
 
 
-def test_discord_ed25519_with_ecdsa_condition():
+def test_discord_json_from_body_string():
```
There is an existing test that leverages ECDSA and JsonCondtion already - see https://github.com/nucypher/nucypher/blob/v7.7.x/tests/unit/conditions/test_sequential_condition.py#L407. Feel free to adjust that test or write a new one for your needs.


Side note: I know this isn't all your code, but any reason most (if not all) of these tests in "integration" can't be moved to the ecdsa "unit" test. I guess the compound condition test is an integration test, but you can move that to the "unit" tests for compound condition i.e. `test_compound_condition.py`. (cc @KPrasch )

derekpierre commented on nucypher/policy/conditions/lingo.py:
```diff
@@ -330,6 +330,18 @@ def __init__(self, operand: Condition):
+        else HexBytes(a.encode() if isinstance(a, str) else str(a).encode()).hex()
+    ),
+    # hashing
+    "keccak": lambda a: HexBytes(keccak(a.encode() if isinstance(a, str) else a)).hex(),
```
Why automatically convert the hash to hex? The user has the option to pair it with the use of "toHex" operation if they want hex.

```suggestion
    "keccak": lambda a: keccak(a.encode() if isinstance(a, str) else a),
```

derekpierre commented on tests/unit/conditions/test_variable_operation.py:
```diff
@@ -62,6 +62,28 @@
+        None,
+        "test",
+        "0x9c22ff5f21f0b81b113e63f7db6da94fedef11b2119b4088b89664fb9a3cb658",
+    ),
```
Let's also add a case for a number and for bytes. Example (but not limited to):
- 24
- b"testing"

derekpierre commented on tests/unit/conditions/test_variable_operation.py:
```diff
@@ -107,9 +129,9 @@ def test_type_errors_in_evaluation(operation):
-    if operation in ["bool", "str"]:
+    # Skip type error test for operations that can handle any input without raising TypeError
+    # or that raise different exceptions (e.g., JSONDecodeError for fromJson)
+    if operation in ["bool", "str", "toJson", "fromJson", "toHex", "fromHex", "keccak"]:
```
Made some recommendations on changing the logic for these new operations logic. Consequently,
- `toJson` (i.e. `json.loads()`) fails if bytes are passed to it
- `fromJson` (i.e. json.dumps()`) fails if bytes are passed to it
- `toHex` fails if an array like that in this test or a float is provided to it
- `fromHex` fails if an array like that in this test or a float is provided to it 
- `keccak` fails if an array like that in this test a float is provided to it.

So you shouldn't need to skip these - try playing around with the above.

derekpierre commented on nucypher/policy/conditions/lingo.py:
```diff
@@ -330,6 +330,18 @@ def __init__(self, operand: Condition):
+    "keccak": lambda a: HexBytes(keccak(a.encode() if isinstance(a, str) else a)).hex(),
 }
 
 MAX_VARIABLE_OPERATIONS = 5
```
Given the complexity of the processing you are now using, do you think a max of 5 is still sufficient? I'm purely curious since 5 was just random, but it can be modified.

theref replied:
i can't remember how it's applied. But i think we'll probably do more than 5 total variable operations in the production CL condition

derekpierre reviewed (dismissed):
Nice work. Recommending some adjustments to the logic for the new operations.

Side note: Do these operations fully solve our problem outlined in https://github.com/nucypher/sprints/issues/271? We've tested this on the test Discord server? Just want to make sure we are being comprehensive here before merging.

theref committed:
Address PR feedback: fix toHex operator and improve test quality

- Fix toHex operator to handle integer types correctly (addresses derekpierre feedback)
  - Extract toHex to _to_hex() helper function for better readability
  - Properly handle int, bytes, bytearray, and str types
  - Raise TypeError for invalid types like float, None, list
- Add comprehensive toHex type error tests
- Improve test comments per Copilot feedback
  - Clarify why operations are excluded from type error tests
  - Split exclusion logic into two clear sections
- Extract duplicate Discord test constants (addresses cygnusv feedback)
  - Add DISCORD_PUBLIC_KEY_HEX, DISCORD_SIGNATURE_HEX, DISCORD_TIMESTAMP, DISCORD_BODY constants
  - Replace hardcoded values in both Discord integration tests

theref committed:
Address Derek's second round of PR feedback

- Simplify fromJson to just json.loads(a) without encoding logic
- Change keccak operator to return bytes instead of hex string
- Update keccak implementation to handle int and other types by converting to string representation
- Add test cases for keccak with int (24) and bytes (b"testing")
- Update all keccak test expectations from hex strings to bytes
- Remove test_keccak_type_errors since keccak now handles all types via string conversion
- Update type error skip list to include keccak as it accepts any type

manumonti reviewed (dismissed):
I would have approved the PR if I hadn't read Derek's concerns on  `toHex`. So I'm going to wait until this is solved.

Looks great to me!

theref commented:
> I would have approved the PR if I hadn't read Derek's concerns on `toHex`. So I'm going to wait until this is solved.
> 
> Looks great to me!

I've already addressed that issue with a new `_to_hex` function

KPrasch commented on tests/integration/conditions/test_ecdsa_condition_integration.py:
```diff
@@ -129,30 +144,257 @@ def test_ecdsa_in_compound_condition():
+    # Create ECDSA condition for signature verification with message template
     ecdsa_condition = ECDSACondition(
-        message=":message",
+        message=":timestamp:discordPayload",  # Constructs message from timestamp + payload
```
I accidentally left this in... I’ll refactor to remove it entirely (unless we decide this type of expression is useful?).

derekpierre commented on nucypher/policy/conditions/lingo.py:
```diff
@@ -330,6 +352,18 @@ def __init__(self, operand: Condition):
+        a.encode()
+        if isinstance(a, str)
+        else (a if isinstance(a, (bytes, bytearray)) else str(a).encode())
+    ),
```
Let's use `keccak()` as is without casting to string since it already handles the relevant types for you. The only reason we need to convert a string to bytes is because if you are passing string to `keccak` you are supposed to use the `text=` kwarg; instead we convert it to bytes which `keccak` would have done internally anyway.

```suggestion
    "keccak": lambda a: keccak(a.encode() if isinstance(a, str) else a)
```

derekpierre commented on tests/unit/conditions/test_variable_operation.py:
```diff
@@ -107,9 +143,17 @@ def test_type_errors_in_evaluation(operation):
+    # These operations are designed to accept any input type and will not raise TypeError.
+    # - bool, str, toJson: explicitly designed to handle any type
+    # - keccak: converts any type to string representation before hashing
+    if operation in ["bool", "str", "toJson", "keccak"]:
```
`toJson` and  `keccak` doesn't actually handle any value...  - see comment here - https://github.com/nucypher/nucypher/pull/3664#discussion_r2472940190.

derekpierre replied:
Fixed via #3667

theref committed:
refactor(conditions): simplify keccak operator to use native type handling

Remove str(a).encode() fallback from keccak operator. Now directly passes
non-string values to eth_utils.keccak, which natively supports bytes,
bytearray, int, and bool types. This simplifies the implementation and
makes type handling more explicit.

theref committed:
test(conditions): fix keccak(int) test case and type error skip lists

- Update keccak(24) expected value to match new implementation that hashes
  byte value 24 directly instead of string "24"
- Fix type error skip lists: move toJson and keccak to appropriate skip
  lists based on their actual type handling behavior
- Clarify skip list comments to accurately describe why operations are skipped

derekpierre commented:
Side note: @theref, if you're wondering "what next", we'll need to add these new operators to `taco-web` as well.

derekpierre committed:
Update variable operation type error to account for new operations: to/fromJson, to/fromHex, keccak.

theref committed:
Merge pull request #3667 from derekpierre/json-hex-tweak

Update variable operation type error to account for new operations

manumonti reviewed (approved):
LGTM! 💫

Muhammad-Altabba reviewed (commented):
I am not very much into python and nucypher, but the PR looks good to me.

derekpierre commented on tests/unit/conditions/test_variable_operation.py:
```diff
@@ -107,11 +143,17 @@ def test_type_errors_in_evaluation(operation):
+    # Skip type error test for operations that may raise exceptions other than TypeError,
+    # such as JSONDecodeError for 'fromJson' or ValueError for 'fromHex'.
+    # Or operations that have their own specific TypeError tests.
+    if operation in ["fromJson", "fromHex", "toHex", "toJson", "keccak"]:
```
@theref - I no longer see the changes from https://github.com/nucypher/nucypher/pull/3667 - was this lost as part of the rebase?

derekpierre replied:
^ Happy to approve once this is addressed/investigated.

derekpierre replied:
FIxed rebase

KPrasch reviewed (approved):
Tested on Devnet ✔️
